### PR TITLE
[21pt] PR: remove V flag and add output resolution flag

### PR DESCRIPTION
**Pull request by AliForghani-NOAA**
This PR removes the V flag and added a new flag to control the spatial resolution of 'simplified' rasters. 

This PR is related to issues #63 and #53 

### Additions  
- A new `src/errors.py` to hold the customized exceptions definitions for the entire program

### Changes  

Common changes among all below files are 1) Removing V flag argument 2) Renaming 'b_is_metric_***' or 'b_is_feet' variables into 'model_unit'. Other changes include:

- `src/clip_dem_from_shape.py` 
&emsp;&emsp; 1. Determined 'model_unit' using 'model_unit_from_crs()' function of shared_functions.py
&emsp;&emsp; 2. Changed the default buffer value from 700 to 300 to be consistent with the hard-coded value used in ras2fim.py

- `src/convert_tif_to_ras_hdf5.py` 
&emsp;&emsp; 1. Determined 'model_unit' using 'model_unit_from_crs()' function of shared_functions.py

- `src/create_fim_rasters.py`
&emsp;&emsp; 1. Improved the arguments description to allow directly running Step 5 using this file
&emsp;&emsp; 2. Removed 'flt_out_resolution' argument which were not being used anywhere in the code. 

- `src/get_usgs_dem_from_shape.py`
&emsp;&emsp; 1. Determined 'model_unit' using 'model_unit_from_crs()' function of shared_functions.py

- `src/ras2fim.py`
&emsp;&emsp; 1. Removed 'vert_unit,' argument and moved all codes to check the vertical unit of the code into shared_functions.py 
&emsp;&emsp; 2. Added the new 'output_resolution' and 'model_unit' arguments
&emsp;&emsp; 3. Removed the hard-coded output resolution (for simplified rasters) and assigned it with ''output_resolution' flag
&emsp;&emsp; 4. Determined the model unit using 1) -p entry and 2) reading all HEC-RAS models prj files-- Used 'confirm_models_unit() function of shared_functions.py

- `src/shared_functions.py`
Added three below functions to determine model units throughout the entire program:
&emsp;&emsp; 1. confirm_models_unit()
&emsp;&emsp; 2. model_unit_from_crs()
&emsp;&emsp; 3. model_unit_from_ras_prj() 

- `src/simplify_fim_rasters.py`
&emsp;&emsp;  Changed the default value of output resolution from 3m to 10m 

- `src/worker_fim_rasters.py`
&emsp;&emsp; 1. Determined model_unit using 'model_unit_from_ras_prj()' function of shared_functions.py

### Testing
For HUC 12090301 (with the unit in feet) successfully tested the code's behavior  with different -p projection flags (with units of ft or meter).

### Checklist

_You may update this checklist before and/or after creating the PR. If you're unsure about any of them, please ask, we're here to help! These items are what we are going to look for before merging your code._

- [x] Informative and human-readable title, using the format: `[_pt] PR: <description>`
- [x] Links are provided if this PR resolves an issue, or depends on another other PR
- [x] If submitting a PR to the `dev` branch (the default branch), you have a descriptive [Feature Branch](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow) name using the format: `dev-<description-of-change>` (e.g.: `dev-revise-levee-masking`)
- [x] Changes are limited to a single goal (no scope creep)
- [x] The feature branch you're submitting as a PR is up to date (merged) with the latest `dev` branch
- [x] Changes adhere to [PEP-8 Style Guidelines](https://peps.python.org/pep-0008/)
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] Project documentation has been updated ([CHANGELOG](/doc/CHANGELOG.md) and/or [README](/README.md))
- [x] [Reviewers requested](https://help.github.com/articles/requesting-a-pull-request-review/)

### CODE REVIEW

**Comment by RobHanna-NOAA** [comment]
While I didn't test it, I did a fast glance at styling and it looks really good. Well done @AliForghani-NOAA . Laura and/or Emily can test this.  Also, loved that you started the first of the 'errors.py' class as we need it and  can start getting it added in tons of places. Thanks for getting that started.

**Review by RobHanna-NOAA — APPROVED** [review]

**Review by EmilyDeardorff — APPROVED** [review]
Runs as expected and excellent additions to the codebase.

**Comment by EmilyDeardorff** [comment]
Merged into my development branch and ran it with a smaller version of ras2fim. I was expecting to see "meters" in the run_arguments.txt file and instead it had "feet" as the model unit (since that is the vertical unit of the terrain rasters I'm using). I'm not sure if the model unit is off or if feet is accurate for some part of the model but not the terrain vertical unit.

image

**Comment by AliForghani-NOAA** [comment]
The “model_unit” shows the unit of HECRAS models. The input/given DEM must
have z unit of meter, regardless of the hec Ras models unit. What is the
vertical unit of the terrain raster you are using?

**Comment by AliForghani-NOAA** [comment]
Rob,
We do see “model_unit” in the run_arguments.txt file. Actually,
“model_unit” is one of the first variable assigned in the code. No action
is needed in that regard.

**Comment by RobHanna-NOAA** [comment]
ah gotcha. Then forgive my ignorance.. that value is the CRS we want for the output tifs? (so we convert the dem to that CRS before we process most of our ras2fim steps?

**Comment by AliForghani-NOAA** [comment]
No worries Rob. 
The "model_unit" simply refers to the measurement unit of HEC-RAS models, which is either meter or feet. This unit is inferred from CRS/EPSG (-p flag). The necessity to have the "model_unit" is mostly for z unit management ( including managing z unit of the given DEM and folder 03 terrain). Spatially, the DEM is reprojected into -p flag CRS/EPSG.  The crs of the output tifs is determined by -p flag, except outputs of Step 6 where we "simplifiy" them.

**Comment by EmilyDeardorff** [comment]
After looking into it further, it seems like older versions of ras2fim displayed a bug where the 03_terrain DEM's had a vertical unit that differed from the HEC-RAS model unit. In this case, if the input DEM was in meters, the 03_terrain DEM was also in meters. (This is shown in the first screenshot, where the larger input DEM and the smaller 03_terrain DEM have the same value, indicating that they are in the same unit.)

I ran the ras2fim script in this PR and the resulting 03_terrain DEM's were in the same unit as the HEC-RAS model unit, which is the expected performance. (This is shown in the second screenshot, where the larger input DEM has a value in meters and the smaller, overlaid DEM from 03_terrain has a value in feet, which is the same as the HEC-RAS model unit.)

In summary, the model unit mismatch issue that I commented on earlier seems to be resolved in this PR.

----------------------

Example 1: 03_terrain DEM is in the same units as the input DEM, which indicates that the DEM was not properly converted during processing
image

Exampple 2: 03_terrain DEM is in feet, whereas the input DEM is in meters, which indicates that the dem *was* properly converted to the HEC-RAS model unit
image

**Review by EmilyDeardorff — APPROVED** [review]
